File: packages/ios-app/Sources/Models/Tokens/TokenStateManager.swift
import Foundation
import os

// MARK: - Token State Manager

/// Manages session-level token tracking state.
///
/// This provides a unified token tracking system that replaces the previous
/// fragmented tracking. It mirrors the agent-side TokenStateManager.
///
/// Key responsibilities:
/// - Recording per-turn token usage
/// - Accumulating totals for billing display
/// - Tracking context window state for progress bar
/// - Maintaining history for audit trail
/// - Supporting session resume/restore
@Observable
@MainActor
final class TokenStateManager {
    private let logger = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "com.tron",
        category: "TokenStateManager"
    )

    // MARK: - State

    /// Current turn's record (most recent)
    private(set) var current: TokenRecord?

    /// Accumulated totals (for billing)
    private(set) var accumulated: AccumulatedTokens

    /// Context window state
    private(set) var contextWindow: ContextWindowState

    /// History of all token records (for audit trail)
    private(set) var history: [TokenRecord] = []

    // MARK: - Initialization

    init(maxContextSize: Int = 1_000_000) {
        self.accumulated = AccumulatedTokens()
        self.contextWindow = ContextWindowState(maxSize: maxContextSize)
    }

    // MARK: - Recording

    /// Update from server's turn_end event with TokenRecord
    func updateFromTurnEnd(_ record: TokenRecord) {
        // Validate: log warning if tokens are 0
        if record.source.rawInputTokens == 0 && record.source.rawOutputTokens == 0 {
            logger.warning("[TOKEN-ANOMALY] Received zero tokens in turn_end (turn=\(record.meta.turn))")
        }

        // Update current
        current = record

        // Accumulate
        accumulated.inputTokens += record.source.rawInputTokens
        accumulated.outputTokens += record.source.rawOutputTokens
        accumulated.cacheReadTokens += record.source.rawCacheReadTokens
        accumulated.cacheCreationTokens += record.source.rawCacheCreationTokens

        // Update context window
        contextWindow.currentSize = record.computed.contextWindowTokens
        updateCalculatedValues()

        // Track history
        history.append(record)

        let contextSize = self.contextWindow.currentSize
        logger.debug("[TOKEN-STATE] Updated from turn_end: turn=\(record.meta.turn) contextWindow=\(contextSize) newInput=\(record.computed.newInputTokens)")
    }

    // MARK: - State Restoration

    /// Restore state from reconstructed events (session resume)
    func restoreFromEvents(_ records: [TokenRecord], accumulated: AccumulatedTokens) {
        self.history = records
        self.current = records.last
        self.accumulated = accumulated

        if let last = records.last {
            contextWindow.currentSize = last.computed.contextWindowTokens
        }
        updateCalculatedValues()

        let contextSize = self.contextWindow.currentSize
        logger.info("[TOKEN-RESTORE] Restored \(records.count) records, contextWindow=\(contextSize)")
    }

    /// Restore accumulated values from session data
    func setAccumulatedTokens(
        inputTokens: Int,
        outputTokens: Int,
        cacheReadTokens: Int,
        cacheCreationTokens: Int,
        cost: Double = 0
    ) {
        accumulated.inputTokens = inputTokens
        accumulated.outputTokens = outputTokens
        accumulated.cacheReadTokens = cacheReadTokens
        accumulated.cacheCreationTokens = cacheCreationTokens
        accumulated.cost = cost
    }

    /// Set context window from reconstructed state
    func setContextWindowSize(_ size: Int) {
        contextWindow.currentSize = size
        updateCalculatedValues()
    }

    // MARK: - Context Limit

    /// Set the maximum context window size (e.g., when model changes)
    func setContextLimit(_ limit: Int) {
        contextWindow.maxSize = limit
        updateCalculatedValues()
    }

    // MARK: - Cost Accumulation

    /// Add cost for a turn
    func addCost(_ cost: Double) {
        accumulated.cost += cost
    }

    // MARK: - Reset

    /// Reset all state for a new session
    func reset() {
        current = nil
        history.removeAll()
        accumulated = AccumulatedTokens()
        contextWindow = ContextWindowState(maxSize: contextWindow.maxSize)
    }

    // MARK: - Private Helpers

    private func updateCalculatedValues() {
        let maxSize = contextWindow.maxSize
        let currentSize = contextWindow.currentSize

        // Calculate percentage (cap at 100)
        let rawPercent = maxSize > 0 ? Double(currentSize) / Double(maxSize) * 100 : 0
        contextWindow.percentUsed = min(100, Int(rawPercent.rounded()))

        // Calculate tokens remaining (floor at 0)
        contextWindow.tokensRemaining = max(0, maxSize - currentSize)
    }
}

// MARK: - Supporting Types

/// Accumulated token totals across all turns in a session.
struct AccumulatedTokens {
    var inputTokens: Int = 0
    var outputTokens: Int = 0
    var cacheReadTokens: Int = 0
    var cacheCreationTokens: Int = 0
    var cost: Double = 0

    var totalTokens: Int { inputTokens + outputTokens }
}

/// Current state of the context window.
struct ContextWindowState {
    var currentSize: Int = 0
    var maxSize: Int
    var percentUsed: Int = 0
    var tokensRemaining: Int

    init(maxSize: Int = 1_000_000) {
        self.maxSize = maxSize
        self.tokensRemaining = maxSize
    }
}

File: packages/agent/src/infrastructure/tokens/state/token-state-manager.ts
/**
 * @fileoverview Token State Manager
 *
 * Manages session-level token tracking state. This is the single source of truth
 * for all token-related data in a session.
 *
 * Key responsibilities:
 * 1. Recording per-turn token usage
 * 2. Accumulating totals for billing
 * 3. Tracking context window state
 * 4. Maintaining history for audit trail
 * 5. Handling provider switches
 * 6. Supporting session resume/restore
 */

import { createLogger } from '@infrastructure/logging/index.js';
import type {
  TokenSource,
  TokenMeta,
  TokenRecord,
  TokenState,
  AccumulatedTokens,
  ProviderType,
} from '../types.js';
import { createEmptyTokenState, createEmptyAccumulatedTokens } from '../types.js';
import { normalizeTokens } from '../normalization/index.js';

const logger = createLogger('token-state-manager');

// =============================================================================
// Configuration
// =============================================================================

/**
 * Configuration for TokenStateManager.
 */
export interface TokenStateManagerConfig {
  /** Maximum context window size for the model (default: 200,000) */
  contextLimit?: number;
}

// =============================================================================
// TokenStateManager Class
// =============================================================================

/**
 * Manages session-level token tracking state.
 *
 * This replaces the previous fragmented token tracking systems with a single,
 * unified state manager that provides:
 * - Immutable TokenRecords for each turn
 * - Accumulated totals for billing
 * - Context window tracking for progress display
 * - Full history for audit trail
 * - Provider-aware normalization
 */
export class TokenStateManager {
  private state: TokenState;
  private currentProviderType: ProviderType = 'anthropic';

  constructor(config: TokenStateManagerConfig = {}) {
    this.state = createEmptyTokenState(config.contextLimit ?? 200_000);
  }

  // ---------------------------------------------------------------------------
  // Recording
  // ---------------------------------------------------------------------------

  /**
   * Record token usage for a turn.
   *
   * @param source - Token source extracted from provider API response
   * @param meta - Metadata for this turn
   * @param cost - Optional cost for this turn
   * @returns The normalized TokenRecord
   */
  recordTurn(source: TokenSource, meta: TokenMeta, cost?: number): TokenRecord {
    // Get previous baseline from current context
    const previousBaseline = this.state.contextWindow.currentSize;

    // Normalize tokens to create immutable record
    const record = normalizeTokens(source, previousBaseline, meta);

    // Update accumulated totals
    this.state.accumulated.inputTokens += source.rawInputTokens;
    this.state.accumulated.outputTokens += source.rawOutputTokens;
    this.state.accumulated.cacheReadTokens += source.rawCacheReadTokens;
    this.state.accumulated.cacheCreationTokens += source.rawCacheCreationTokens;
    this.state.accumulated.cacheCreation5mTokens += source.rawCacheCreation5mTokens;
    this.state.accumulated.cacheCreation1hTokens += source.rawCacheCreation1hTokens;
    this.state.accumulated.cost += cost ?? 0;

    // Update context window state
    this.state.contextWindow.currentSize = record.computed.contextWindowTokens;
    this.updateContextWindowCalculations();

    // Update current and history
    this.state.current = record;
    this.state.history.push(record);

    logger.info('[TOKEN-STATE] Turn recorded', {
      turn: meta.turn,
      provider: source.provider,
      rawInput: source.rawInputTokens,
      rawOutput: source.rawOutputTokens,
      rawCacheRead: source.rawCacheReadTokens,
      rawCacheCreation: source.rawCacheCreationTokens,
      rawCacheCreation5m: source.rawCacheCreation5mTokens,
      rawCacheCreation1h: source.rawCacheCreation1hTokens,
      contextWindow: record.computed.contextWindowTokens,
      newInput: record.computed.newInputTokens,
      accumulatedInput: this.state.accumulated.inputTokens,
      accumulatedCacheRead: this.state.accumulated.cacheReadTokens,
    });

    return record;
  }

  // ---------------------------------------------------------------------------
  // State Access
  // ---------------------------------------------------------------------------

  /**
   * Get the current token state.
   *
   * Returns a copy to prevent external mutation.
   */
  getState(): Readonly<TokenState> {
    return this.state;
  }

  // ---------------------------------------------------------------------------
  // Provider Management
  // ---------------------------------------------------------------------------

  /**
   * Handle provider type change.
   *
   * When provider changes, we reset the context baseline because different
   * providers interpret inputTokens differently. The previous baseline
   * would be meaningless after a provider switch.
   *
   * Note: We preserve accumulated tokens and history.
   */
  onProviderChange(newProviderType: ProviderType): void {
    if (this.currentProviderType !== newProviderType) {
      logger.info('[TOKEN-STATE] Provider changed', {
        from: this.currentProviderType,
        to: newProviderType,
      });
      this.currentProviderType = newProviderType;
      // Reset context baseline for fresh delta calculation
      this.state.contextWindow.currentSize = 0;
      this.updateContextWindowCalculations();
    }
  }

  // ---------------------------------------------------------------------------
  // Context Limit Management
  // ---------------------------------------------------------------------------

  /**
   * Set the maximum context window size.
   *
   * This should be called when the model changes or when we get
   * updated model information from the server.
   */
  setContextLimit(limit: number): void {
    this.state.contextWindow.maxSize = limit;
    this.updateContextWindowCalculations();
  }

  // ---------------------------------------------------------------------------
  // Session Resume
  // ---------------------------------------------------------------------------

  /**
   * Restore state from previous session data.
   *
   * Used when resuming a session to restore token tracking state.
   *
   * @param data - Previous session data to restore
   */
  restoreState(data: { history: TokenRecord[]; accumulated?: AccumulatedTokens }): void {
    // Restore history
    this.state.history = data.history ?? [];

    // Restore current from last record
    const lastRecord = this.state.history[this.state.history.length - 1];
    this.state.current = lastRecord ?? null;

    // Restore accumulated or use defaults
    this.state.accumulated = data.accumulated ?? createEmptyAccumulatedTokens();

    // Restore context window from current record
    if (this.state.current) {
      this.state.contextWindow.currentSize = this.state.current.computed.contextWindowTokens;
    }
    this.updateContextWindowCalculations();

    logger.info('[TOKEN-RESTORE] State restored', {
      historyCount: this.state.history.length,
      contextWindow: this.state.contextWindow.currentSize,
      accumulatedInput: this.state.accumulated.inputTokens,
    });
  }

  // ---------------------------------------------------------------------------
  // Private Helpers
  // ---------------------------------------------------------------------------

  /**
   * Update derived context window calculations.
   */
  private updateContextWindowCalculations(): void {
    const { currentSize, maxSize } = this.state.contextWindow;

    // Calculate percentage (cap at 100)
    const rawPercent = maxSize > 0 ? (currentSize / maxSize) * 100 : 0;
    this.state.contextWindow.percentUsed = Math.min(100, rawPercent);

    // Calculate tokens remaining (floor at 0)
    this.state.contextWindow.tokensRemaining = Math.max(0, maxSize - currentSize);
  }
}

// =============================================================================
// Factory Function
// =============================================================================

/**
 * Create a new TokenStateManager instance.
 */
export function createTokenStateManager(config: TokenStateManagerConfig = {}): TokenStateManager {
  return new TokenStateManager(config);
}
